### PR TITLE
Refactor/convert to typescript3

### PR DESCRIPTION
**Pull request by creisle**
- Convert to typescript
- Update parser to use v2

### CODE REVIEW

File: test/index.test.ts [hunk 1]
```diff
@@ -130,17 +126,12 @@ describe('SCHEMA', () => {
...
+                break1Start: {
+                    '@class': 'ProteinPosition', pos: 1, refAA: 'A', prefix: 'p',
+                },
                 break1Repr: 'bad',
             }, { addDefaults: true });
             expect(formatted).toHaveProperty('break1Repr', 'p.A1');
         });
     });
-
-    describe('previews and identifiers', () => {
```
Comment (areisle) [inline test/index.test.ts:140]: why did you delete this test?

File: config/jest.config.js [hunk 2]
```diff
@@ -29,13 +30,20 @@ module.exports = {
...
     testPathIgnorePatterns: [
         '/node_modules/',
     ],
     moduleFileExtensions: [
         'js',
+        'ts',
         'json',
     ],
+    // temporarily prevent type checking until finished setting up
+    globals: {
```
Comment (areisle) [inline config/jest.config.js:44]: this should be removed now that the types are finished

**Review by areisle — CHANGES_REQUESTED** [review]

File: test/index.test.ts (hunk 1 above)
Comment (creisle) [reply test/index.test.ts:140]: Because identifiers is a property that we no longer add/user on the schema model. This was leftover and should have been deleted previously

File: src/property.ts [hunk 3]
```diff
@@ -1,12 +1,47 @@
...
+    readonly type: DbType;
+    readonly pattern?: string;
+    readonly description?: string;
+    readonly generated?: boolean;
+    readonly mandatory?: boolean;
+    readonly nullable?: boolean;
+    readonly readOnly?: boolean;
+    readonly generateDefault?: (rec?: unknown) => unknown;
+    readonly check?: (rec?: unknown) => boolean;
+    readonly default: unknown;
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/property.ts:27]: set to optional

File: config/jest.config.js (hunk 2 above)
Comment (creisle) [reply config/jest.config.js:44]: removed

File: src/property.ts [hunk 4]
```diff
@@ -115,9 +128,9 @@ class Property {
                     this.cast = util.castNullableLink;
                 }
             }
-        }
-        if (this.choices && this.cast) {
-            this.choices = this.choices.map(choice => this.cast(choice));
+        } else if (this.choices) {
+            const castFunc: (arg: unknown) => any = this.cast;
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/property.ts:132]: yup, I couldn't get the types to work otherwise

File: src/property.ts [hunk 5]
```diff
@@ -1,12 +1,47 @@
...
+    DbType, PropertyType, ModelType, PropertyTypeDefinition,
+} from './types';
 
+export interface PropertyTypeInput extends Omit<PropertyTypeDefinition, 'linkedClass'> {
+    linkedClass?: ModelType;
+}
+
+export class Property implements PropertyType {
+    readonly name: string;
+    readonly cast?: (value: any) => any;
```
[the remark this answers is not in the document]
Comment (creisle) [reply src/property.ts:17]: When i change that to unknown i get this error later in the constructor which I am not sure how to handle. Any ideas?

image

[the remark this answers is not in the document]
Comment (areisle) [reply src/property.ts:17]: hmm, I guess either `castToRID` should be changed to accept `unknown` as `arg` or can leave `cast` slightly unsafe as `cast?: (value: any) => unknown;`

File: src/property.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (areisle) [reply src/property.ts:132]: if `this.cast` can't be undefined by this point, then type for cast may need to be changed to reflect that. if it can, then I would just change the condition to be `} else if (this.choices && this.cast) {`

if it can be undefined but you know it's not in this particular situation for some reason and don't want a redundant check, then you could do
```ts
this.choices = this.choices.map((choice) => (this.cast as NonNullable<typeof this.cast>)(choice));
```

File: src/property.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (creisle) [reply src/property.ts]: I went with the latter since castToRID is used in several other places

File: src/property.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (creisle) [reply src/property.ts]: done

**Review by areisle — APPROVED** [review]
